crypto/dilithium/p_dilithium_test.cc

```diff
@@ -0,0 +1,289 @@
+
+  buf = (uint8_t *)OPENSSL_malloc(buf_size);
+  ASSERT_NE(buf, nullptr);
+  EXPECT_TRUE(EVP_PKEY_get_raw_public_key(dilithium_pkey, buf, &buf_size));
+
+  buf_size = 0;
+  EXPECT_FALSE(EVP_PKEY_get_raw_public_key(dilithium_pkey, buf, &buf_size));
+
+  uint32_t err = ERR_get_error();
+  EXPECT_EQ(ERR_LIB_EVP, ERR_GET_LIB(err));
+  EXPECT_EQ(EVP_R_BUFFER_TOO_SMALL, ERR_GET_REASON(err));
+  OPENSSL_free(buf);
+  buf = nullptr;
+
+  // Extract private key and check it is of the correct size
+  EXPECT_TRUE(EVP_PKEY_get_raw_private_key(dilithium_pkey, buf, &buf_size));
+  EXPECT_EQ((size_t)DILITHIUM3_PRIVATE_KEY_BYTES, buf_size);
+
+  buf = (uint8_t *)OPENSSL_malloc(buf_size);
+  ASSERT_NE(buf, nullptr);
+  EXPECT_TRUE(EVP_PKEY_get_raw_private_key(dilithium_pkey, buf, &buf_size));
+
+  buf_size = 0;
+  EXPECT_FALSE(EVP_PKEY_get_raw_private_key(dilithium_pkey, buf, &buf_size));
+  err = ERR_get_error();
+  EXPECT_EQ(ERR_LIB_EVP, ERR_GET_LIB(err));
+  EXPECT_EQ(EVP_R_BUFFER_TOO_SMALL, ERR_GET_REASON(err));
+  OPENSSL_free(buf);
+
+  EVP_PKEY_CTX_free(dilithium_pkey_ctx);
+  EVP_PKEY_free(dilithium_pkey);
+}
+
+TEST(Dilithium3Test, KeyComparison) {
+  // Generate two Dilithium3 keys are check that they are not equal.
+  EVP_PKEY_CTX *dilithium_pkey_ctx1 = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
+  ASSERT_NE(dilithium_pkey_ctx1, nullptr);
+
+  EVP_PKEY *dilithium_pkey1 = EVP_PKEY_new();
+  ASSERT_NE(dilithium_pkey1, nullptr);
+
+  EXPECT_TRUE(EVP_PKEY_keygen_init(dilithium_pkey_ctx1));
+  EXPECT_TRUE(EVP_PKEY_keygen(dilithium_pkey_ctx1, &dilithium_pkey1));
+  ASSERT_NE(dilithium_pkey1->pkey.ptr, nullptr);
+
+  EVP_PKEY_CTX *dilithium_pkey_ctx2 = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
+  ASSERT_NE(dilithium_pkey_ctx2, nullptr);
+
+  EVP_PKEY *dilithium_pkey2 = EVP_PKEY_new();
+  ASSERT_NE(dilithium_pkey2, nullptr);
+
+  EXPECT_TRUE(EVP_PKEY_keygen_init(dilithium_pkey_ctx2));
+  EXPECT_TRUE(EVP_PKEY_keygen(dilithium_pkey_ctx2, &dilithium_pkey2));
+  ASSERT_NE(dilithium_pkey2->pkey.ptr, nullptr);
+
+  EXPECT_EQ(0, EVP_PKEY_cmp(dilithium_pkey1, dilithium_pkey2));
+
+  EVP_PKEY_free(dilithium_pkey1);
+  EVP_PKEY_free(dilithium_pkey2);
+  EVP_PKEY_CTX_free(dilithium_pkey_ctx1);
+  EVP_PKEY_CTX_free(dilithium_pkey_ctx2);
+}
+
+TEST(Dilithium3Test, NewKeyFromBytes) {
+  // Test the generation of a Dilithium3 key from bytes
+  // Source key
+  EVP_PKEY_CTX *dilithium_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
+  ASSERT_NE(dilithium_pkey_ctx, nullptr);
+
+  EVP_PKEY *dilithium_pkey = EVP_PKEY_new();
+  ASSERT_NE(dilithium_pkey, nullptr);
+
+  EXPECT_TRUE(EVP_PKEY_keygen_init(dilithium_pkey_ctx));
+  EXPECT_TRUE(EVP_PKEY_keygen(dilithium_pkey_ctx, &dilithium_pkey));
+  ASSERT_NE(dilithium_pkey->pkey.ptr, nullptr);
+  const DILITHIUM3_KEY *dilithium3Key = (DILITHIUM3_KEY *)(dilithium_pkey->pkey.ptr);
+
+  // New raw public key
+  EVP_PKEY *new_public = EVP_PKEY_new_raw_public_key(EVP_PKEY_DILITHIUM3,
```

Comment:
@[andrewhop](https://github.com/andrewhop) andrewhop [on Oct 4](https://github.com/awslabs/aws-lc/pull/637/files#r987430623)
Instead of directly accessing the raw key through the pointer should this test the ASN1 encoding and decoding with EVP_marshal_public_key and EVP_parse_public_key?

Reply:
This test is specifically targeting `EVP_PKEY_new_raw_public_key `and `EVP_PKEY_new_raw_private_key`. But I agree that we should test `EVP_marshal_public_key` and `EVP_parse_public_key` as well. [See the newly added marshal/parse tests here](https://github.com/awslabs/aws-lc/blob/47415745b8e74604e182b38ee24fdd01d3d8ad5e/crypto/dilithium/p_dilithium_test.cc#L194-L222).